vervet-underground/internal/storage/mem/mem.go

```diff
@@ -6,81 +6,274 @@ package mem
+		s.serviceVersionMappedRevisionSpecs[name] = versionMappedRevisionSpecs{}
+	}
+
+	revisions, ok := s.serviceVersionMappedRevisionSpecs[name][version]
+	if ok {
+		if _, ok := revisions[digest]; ok {
 			return nil
 		}
 	} else {
-		s.contents[k] = map[string][]byte{}
+		s.serviceVersionMappedRevisionSpecs[name][version] = mappedRevisionSpecs{}
+		revisions = s.serviceVersionMappedRevisionSpecs[name][version]
+	}
+
+	// If the version is newly initialized, meaning no revisions exist,
+	// create the new vervet.VersionSlice with that version initialized
+	// else, append it to the existing service's VersionSlice
+	if len(revisions) == 0 {
+		if _, ok = s.serviceVersions[name]; !ok {
+			s.serviceVersions[name] = vervet.VersionSlice{*parsedVersion}
+		} else {
+			s.serviceVersions[name] = append(s.serviceVersions[name], *parsedVersion)
+			// sort versions when new ones are introduced to maintain BST functionality
+			sort.Sort(s.serviceVersions[name])
+		}
+	}
+	// End of initializations
+
+	// add the new ContentRevision
+	s.serviceVersionMappedRevisionSpecs[name][version][digest] = ContentRevision{
+		serviceVersion: name + "_" + version,
+		timestamp:      scrapeTime,
+		digest:         digest,
+		blob:           contents,
+	}
+
+	return nil
+}
+
+// CollateVersions does the following:
+//   - calls updateCollatedVersions for a slice of unique vervet.Version entries
+//   - for each unique vervet.Version, run collateVersion to create a compiled VU openapi doc
+func (s *Storage) CollateVersions() error {
+	var errs error
+	s.updateCollatedVersions()
+	for _, version := range s.collatedVersions {
+		err := s.collateVersion(version)
+		if err != nil {
+			log.Error().Err(err).Msgf("Could not Collate Version %s", version.String())
+			errs = multierr.Append(errs, err)
+		}
+	}
+	return errs
+}
+
+func (s *Storage) GetCollatedVersionSpecs() (map[string][]byte, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	versionSpecs := map[string][]byte{}
+	for key, value := range s.collatedVersionedSpecs {
+		json, err := value.MarshalJSON()
+		if err != nil {
+			log.Error().Err(err).Msgf("Could not Marshal Version %s", key.String())
+			return nil, err
+		}
+		versionSpecs[key.String()] = json
+	}
+	return versionSpecs, nil
+}
+
+// updateCollatedVersions collects all unique vervet.Versions detected for every service published using
+func (s *Storage) updateCollatedVersions() {
+	uniqueVersions := make(map[string]vervet.Version)
+
+	// map keys automatically make versions unique rather than iterate through array for each entry
+	// service, versionSlice
+	for _, versionSlice := range s.serviceVersions {
+		// index, version
+		for _, version := range versionSlice {
+			uniqueVersions[version.String()] = version
+		}
+	}
+
+	s.mu.Lock()
+	s.collatedVersions = make([]vervet.Version, len(uniqueVersions))
+	i := 0
+	for _, value := range uniqueVersions {
+		s.collatedVersions[i] = value
+		i++
+	}
```

Comment:
Slightly more compact:

```suggestion
	for i, value := range uniqueVersions {
		s.collatedVersions[i] = value
	}
```

Reply:
`uniqueVersions` is a map. The mechanism is to use the map for creating uniqueness in the array, and then inserting all the values and presizing the array.

I could simplify the code to just use `append` instead if you'd rather but this is slightly more memory efficient